File: android/app/src/main/java/sol/vortexx/android/contacts/impl/HttpContacts.kt
package sol.vortexx.android.contacts.impl

import io.ktor.client.call.body
import io.ktor.client.request.*
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import kotlinx.serialization.Serializable
import sol.vortexx.android.contacts.api.Contact
import sol.vortexx.android.contacts.api.Contacts
import sol.vortexx.android.net.impl.VortexHttpClient
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class HttpContacts @Inject constructor(
    private val http: VortexHttpClient,
) : Contacts {
    @Serializable private data class Wrap(val contacts: List<Contact>)
    @Serializable private data class AddBody(val username: String)

    override suspend fun list(): List<Contact> = runCatching {
        val resp = http.client.get("api/contacts")
        if (!resp.status.isSuccess()) emptyList() else resp.body<Wrap>().contacts
    }.getOrDefault(emptyList())

    override suspend fun add(username: String): Contact? = runCatching {
        val resp = http.client.post("api/contacts") {
            contentType(ContentType.Application.Json)
            setBody(AddBody(username))
        }
        if (!resp.status.isSuccess()) null else resp.body<Contact>()
    }.getOrNull()

    override suspend fun remove(id: Long) {
        runCatching { http.client.delete("api/contacts/$id") }
    }

    override suspend fun search(query: String): List<Contact> = runCatching {
        val resp = http.client.get("api/contacts/search") { url { parameters.append("q", query) } }
        if (!resp.status.isSuccess()) emptyList() else resp.body<Wrap>().contacts
    }.getOrDefault(emptyList())
}

File: ios/Modules/Sources/Contacts/impl/HttpContacts.swift
import Foundation
import Net

public actor HttpContacts: Contacts {
    private let http: HttpClient
    public init(http: HttpClient) { self.http = http }

    private struct Wrap: Codable { let contacts: [Contact] }
    private struct AddBody: Codable { let username: String }

    public func list() async -> [Contact] {
        let req = HttpRequest.get("/api/contacts")
        return (try? await http.send(req, Wrap.self).contacts) ?? []
    }

    public func add(username: String) async -> Contact? {
        guard let req = try? HttpRequest.postJson("/api/contacts", body: AddBody(username: username)) else { return nil }
        return try? await http.send(req, Contact.self)
    }

    public func remove(id: Int64) async {
        let req = HttpRequest.delete("/api/contacts/\(id)")
        _ = try? await http.send(req)
    }

    public func search(_ query: String) async -> [Contact] {
        let req = HttpRequest.get("/api/contacts/search", query: [URLQueryItem(name: "q", value: query)])
        return (try? await http.send(req, Wrap.self).contacts) ?? []
    }
}
